### PR TITLE
Add a function to delete a study object

### LINKED ISSUE #556: Add `optuna.study.delete_study()`.
Thanks to https://github.com/pfnet/optuna/pull/524, users can delete studies using command-line interface as follows:

```bash
optuna delete-study --study-name my-study --storage sqlite:///example.db
```

If we implement `optuna.study.delete_study(study_name, storage)` like `optuna.study.create_study()`, it will be useful for users who embed Optuna into their software.

### PR DESCRIPTION
**Pull request by c-bata**
to close #556

### CODE REVIEW

File: tests/test_study.py [hunk 1]
```diff
@@ -539,6 +539,21 @@ def test_load_study(storage_mode, cache_mode):
...
+    # type: (str, bool) -> None
+
+    with StorageSupplier(storage_mode, cache_mode) as storage:
+        # Test deleting a non-existing study.
+        with pytest.raises(ValueError):
+            optuna.delete_study("invalid-study-name", storage)
+
+        # Test deleting an existing study.
+        study = optuna.create_study(storage=storage, load_if_exists=False)
+        optuna.delete_study(study.study_name, storage)
```
Comment (sile) [inline tests/test_study.py]: I think that it's better to add the following code to check that the study has been deleted (unless the class of `storage` is `InMemoryStorage`):
```python
with pytest.raises(ValueError):
    optuna.delete_study(study.study_name, storage)
```

Comment (c-bata) [reply tests/test_study.py]: Can I use `if` statement to skip InMemoryStorage? Some people who like xUnit Test Patterns might say it's an anti-pattern.

Comment (c-bata) [reply tests/test_study.py]: (xUnit Test Patterns is one of the most famous book which introduce a lot of anti-patterns on unit tests)

> Conditional Test Logic is one factor that makes tests more complicated than they really should be.
> http://xunitpatterns.com/Conditional%20Test%20Logic.html

Comment (sile) [reply tests/test_study.py]: > If you don't mind it, I'll add the assertion in this method. If you mind it, I'll separate test methods.

I think that either is OK.
FYI. We sometimes use `if` for such purpose (e.g., [test_study.py#L507](https://github.com/pfnet/optuna/blob/master/tests/test_study.py#L507)).

Comment (c-bata) [reply tests/test_study.py]: Thank you for your review. I fixed.

Comment (sile) [reply tests/test_study.py]: Thank you!

File: optuna/study.py [hunk 2]
```diff
@@ -677,6 +677,27 @@ def load_study(
     return Study(study_name=study_name, storage=storage, sampler=sampler, pruner=pruner)
 
 
+def delete_study(
+        study_name,  # type: str
+        storage=None,  # type: Union[None, str, storages.BaseStorage]
```
Comment (sile) [inline optuna/study.py]: IMO, `storage=None` is meaningless for this method (it'll create an empty study and delete it immediately).
How about removing the default value of `storage` and modifying the type hint to `Union[str, storages.BaseStorage]`? > @c-bata @toshihikoyanase @g-votte

Comment (c-bata) [reply optuna/study.py]: Agree. Does it apply to `create_study` and `load_study`?

Comment (toshihikoyanase) [reply optuna/study.py]: I agree with @sile.
I think it is similar to `storage` argument of [`load_study`](https://optuna.readthedocs.io/en/stable/reference/study.html#optuna.study.load_study).

Comment (sile) [reply optuna/study.py]: I think that the default storage argument of `create_study` (i.e., `create_study(storage=None)`) is useful to create an empty in-memory storage.

Comment (c-bata) [reply optuna/study.py]: Thank you. I applied your feedbacks and pushed it.

Comment (sile) [reply optuna/study.py]: Hmm, `storage` still has the default value: https://github.com/pfnet/optuna/pull/558/files#diff-4e020643693eb296936f5dd7194bd6d6R682

Comment (c-bata) [reply optuna/study.py]: Oh, I missed!

File: optuna/study.py [hunk 3]
```diff
@@ -677,6 +677,27 @@ def load_study(
...
+        storage,  # type: Union[str, storages.BaseStorage]
+):
+    # type: (...) -> None
+    """Delete a :class:`~optuna.study.Study` object.
+
+    Args:
+        study_name:
+            Study's name.
+        storage:
+            Database URL. If this argument is set to None, in-memory storage is used, and the
```
Comment (sile) [inline optuna/study.py]: Please update this document because `storage` doesn't take `None` anymore.

File: tests/test_study.py [hunk 4]
```diff
@@ -539,6 +539,31 @@ def test_load_study(storage_mode, cache_mode):
         assert created_study.study_id == loaded_study.study_id
 
 
+@pytest.mark.parametrize('storage_mode', STORAGE_MODES)
+@pytest.mark.parametrize('cache_mode', CACHE_MODES)
+def test_delete_study(storage_mode, cache_mode):
+    # type: (str, bool) -> None
+
+    with StorageSupplier(storage_mode, cache_mode) as storage:
+        # Get storage object because delete_study does not accept None
```
Comment (toshihikoyanase) [inline tests/test_study.py]: Please add a period at the end of the sentence.

**Comment by toshihikoyanase** [comment]
@c-bata 
We've fixed the issue on TensorFlow v2 by this PR (https://github.com/pfnet/optuna/pull/562). Please reflect the changes by merging the master branch into yours.

**Comment by c-bata** [comment]
Thank you! I pushed the merge commit.

**Pull request merged**